lib/vega_lite/data.ex

```diff
@@ -0,0 +1,231 @@
+    build_fields(fields, root, cols)
+  end
+
+  @doc """
+  Returns the specification of a heat map for a given data and a list of fields to be encoded.
+
+  As a specialized chart, the heatmap expects an `:x` and `:y` and optionally a `:color` and a
+  `:text` field. Defaults to `:nominal` for the axes and `:quantitative` for color and text if
+  types are not specified.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
+
+  """
+  @spec heatmap(Table.Reader.t(), keyword()) :: VegaLite.t()
+  def heatmap(data, fields), do: heatmap(Vl.new(), data, fields)
+
+  @doc """
+  Same as heatmap/2, but takes a valid `VegaLite` specification as the first argument.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Vl.new(title: "Heatmap", width: 500)
+      |> Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
+  """
+  @spec heatmap(VegaLite.t(), Table.Reader.t(), keyword()) :: VegaLite.t()
+  def heatmap(vl, data, fields) do
+    fields = standardize_fields(fields) |> Enum.map(&heatmap_defaults/1)
+
+    case {fields[:x], fields[:y], fields[:text]} do
+      {nil, _, _} -> raise ArgumentError, "the x axis is required to plot a heatmap"
+      {_, nil, _} -> raise ArgumentError, "the y axis is required to plot a heatmap"
+      {_, _, nil} -> chart(vl, data, :rect, fields)
+      _ -> annotated_heatmap(vl, data, fields)
+    end
+  end
+
+  defp annotated_heatmap(vl, data, fields) do
+    text_fields = [text: fields[:text], x: fields[:x], y: fields[:y]]
+    fields = Keyword.delete(fields, :text)
+    used_fields = fields |> Keyword.values() |> used_fields()
+
+    vl
+    |> Vl.data_from_values(data, only: used_fields)
+    |> Vl.layers([layer(data, :rect, fields), layer(data, :text, text_fields)])
+  end
+
+  defp heatmap_defaults({field, opts}) when field in [:x, :y] do
+    {field, Keyword.put_new(opts, :type, :nominal)}
+  end
+
+  defp heatmap_defaults({field, opts}) when field in [:color, :text] do
+    {field, Keyword.put_new(opts, :type, :quantitative)}
+  end
+
+  defp encode_mark(vl, opts) when is_list(opts) do
+    {mark, opts} = Keyword.pop!(opts, :type)
+    Vl.mark(vl, mark, opts)
+  end
+
+  defp encode_mark(vl, mark), do: Vl.mark(vl, mark)
+
+  defp encode_field(schema, cols, field, opts) do
+    {col, opts} = Keyword.pop!(opts, :field)
+    opts = Keyword.put_new_lazy(opts, :type, fn -> cols[col] end)
+    Vl.encode_field(schema, field, col, opts)
+  end
+
+  defp layer(data, mark, fields) do
+    cols = columns_for(data)
+    root = Vl.new() |> encode_mark(mark)
+    build_fields(fields, root, cols)
+  end
+
+  defp used_fields(fields) do
+    for field <- fields, field, uniq: true do
```

Comment:
When would `field` be nil or false?

Reply:
Never and I think I missed the point of the question 🤔